blog/content/post/2022-12-22-fobnail_application.md

```diff
@@ -0,0 +1,363 @@
+- commands written in rules are executed by `sh`
+  - some built-ins behave slightly differently
+  - usually easier to just call a script with different shell (e.g. `bash`)
+    specified in shebang
+  - `at` also executes command in `sh`
+
+For these reasons we will use very simple wrapper scripts (example for mount.sh,
+umount.sh will also need similar wrapper):
+
+```bash
+#!/bin/bash
+
+echo /full/path/to/mount.sh | at -M now
+```
+
+Of course, use your own path to the script. Save it as `mount_wrap.sh`. Now all
+that's left to do is starting this scripts on connection and disconnection of
+the Token. We can do so by creating `/etc/udev/rules.d/99-fobnail.rules` with
+following content:
+
+```
+ACTION=="add", SUBSYSTEM=="net", ENV{ID_VENDOR_ID}=="1234", ENV{ID_MODEL_ID}=="4321" RUN+="/full/path/to/mount_wrap.sh"
+ACTION=="remove", SUBSYSTEM=="net", ENV{ID_VENDOR_ID}=="1234", ENV{ID_MODEL_ID}=="4321" RUN+="/full/path/to/umount_wrap.sh"
+```
+
+Again, use proper paths to scripts. Each rule must be written in one line, they
+can't be broken with backslash. `1234` and `4321` (hexadecimal) are vendor and
+device IDs [used by Fobnail](https://github.com/fobnail/fobnail/blob/main/pal/pal_nrf/src/usb/mod.rs#L15),
+respectively. These may change if we decide to register as USB vendor.
+
+#### Use
+
+Just plug in the Token, wait few seconds (up to few dozen) for green light and
+voila! You should now have access to all your secrets. When the Token is pulled
+out, the secure disk should be automatically unmounted.
+
+## Demo
+
+**TBD**
+
+## Known issues
+
+Attester assumes full control over TPM while it's running. This means that no
+other application can use TPM together with Attester, including another instance
+of Attester. `timeout` is used to work around potential issue with Attester, for
+example when it can't access the Token and gets stuck.
+
+Due to complicated ordering done by `systemd`, Fobnail Token sometimes doesn't
+have its IP assigned properly, `NetworkManager` complicates things even more.
+This can be checked with `networkctl list fobnail` when Token is plugged in -
+`routable` is expected state, but sometimes it is `carrier` instead. The latter
+case won't work, but usually it fixes itself by unplugging and plugging the
+token again.
+
+`systemd` configuration presented above blindly renames `usb0` network interface
+to `fobnail`. This won't work if you're using USB network card. Even though
+[manual for .link files](https://www.freedesktop.org/software/systemd/man/systemd.link.html)
+linked on [official systemd documentation page](https://systemd.io/PREDICTABLE_INTERFACE_NAMES/)
+says that we can use `Property=` in `Match` section to check for particular
+vendor and device ID, this is not true. Trying to do so results in a warning
+being printed to log and **every interface** being matched by this rule,
+including the one used to connect to the Internet and even loopback device `lo`.
```

Comment:
@TomaszAIR @macpijan I guess we probably solved that issue already for embedded devices?

Reply:
@pietrushnic in some projects we assinged name based on pci bus ID. We needed to make sure there that link files will be matched only for eth and en ports by changing `[Match]` section in `/lib/systemd/network/99-default.link` to

```
[Match]
OriginalName=en* eth*
```

But this look like another workaround. I do not remember using `Property=` option.

Reply:
@TomaszAIR thank, anyway, we should have knowledge about how to solve the issue in systemd.

Reply:
I also changed name based on bus/dev/port on my PC, but for USB it means that I have to plug it always to the same slot. It was enough for my needs, but definitely isn't user friendly.

Reply:
@macpijan we have to address that later in the project. Please let me know you are aware of the issues and we will schedule solving that. If we agree scope of this item is clear, this conversation can be resolved.

Reply:
It appears that the problem was caused by old systemd on my system, it works in most recent OSes. Information about it is included in the post.